Check for code usage fluctuations in native images

zakkak opened the pull request:
This aims to aid in detecting code usage fluctuations as early as
possible. I set the threshold to 3% which might be a bit tight.

Ideally most (if not all) ITs being tested in native mode should include
such checks. This way when a dependency gets updated and brings in more
bloat, or a GraalVM change results in more code becoming reachable we
will be able to notice.

If significant fluctuations between GraalVM/Mandrel versions are
detected the annotations `@DisableIfBuiltWithGraalVMOlderThan` and
`@DisableIfBuiltWithGraalVMNewerThan` may be used to run tests with
different properties files.

The properties files can be generated by running the following script,
passing the build output json file as the first parameter and the path
to the target properties file as the second parameter (it will be
overwritten), after a failed run. This way enabling the check for a new
integration test is a matter of:

1. copying `ImageMetricsITCase.java` in the corresponding folder
2. running the test once
3. running the following script passing the
build output json file as the first parameter.

```bash
#!/bin/bash

KEYS=( \
image_details.total_bytes \
analysis_results.types.reachable \
analysis_results.methods.reachable \
analysis_results.fields.reachable \
analysis_results.types.reflection \
analysis_results.methods.reflection \
analysis_results.fields.reflection \
analysis_results.types.jni \
analysis_results.methods.jni \
analysis_results.fields.jni \
)

echo "# Properties file used by ImageMetricsITCase" > $2
for i in "${KEYS[@]}"
do
  echo "$i=$(jq .$i $1)" >> $2
  echo "$i.tolerance=3" >> $2
done
```

Implements part of #36066

Supersedes https://github.com/quarkusio/quarkus/pull/36091

After a first round of feedback I can add the check to more integration tests.

zakkak committed:
Check for code usage fluctuations in native images

This aims to aid in detecting code usage fluctuations as early as
possible. I set the threshold to 3% which might be a bit tight.

Ideally most (if not all) ITs being tested in native mode should include
such checks. This way when a dependency gets updated and brings in more
bloat, or a GraalVM change results in more code becoming reachable we
will be able to notice.

If significant fluctuations between GraalVM/Mandrel versions are
detected the annotations `@DisableIfBuiltWithGraalVMOlderThan` and
`@DisableIfBuiltWithGraalVMNewerThan` may be used to run tests with
different properties files.

The properties files can be generated by running the following script,
passing the build output json file as the first parameter and the path
to the target properties file as the second parameter (it will be
overwritten), after a failed run. This way enabling the check for a new
integration test is a matter of:

1. copying `ImageMetricsITCase.java` in the corresponding folder
2. running the test once 3. running the following script passing the
build output json file as the first parameter.

```bash

KEYS=( \
image_details.total_bytes \
analysis_results.types.reachable \
analysis_results.methods.reachable \
analysis_results.fields.reachable \
analysis_results.types.reflection \
analysis_results.methods.reflection \
analysis_results.fields.reflection \
analysis_results.types.jni \
analysis_results.methods.jni \
analysis_results.fields.jni \
)

echo "# Properties file used by ImageMetricsITCase" > $2
for i in $KEYS
do
  echo "$i=$(jq .$i $1) >> $2
  echo "$i.tolerance=3 >> $2
done
```

zakkak commented:
cc @jerboaa

Sanne commented on integration-tests/jpa-postgresql-withxml/src/test/java/io/quarkus/it/jpa/postgresql/ImageMetricsITCase.java:
```diff
@@ -0,0 +1,80 @@
+import org.junit.jupiter.api.Test;
+
+@QuarkusIntegrationTest
+public class ImageMetricsITCase {
```
Looks like you expect people to copy/paste this class in various modules, then edit the constants.
Perhaps it could be reworked to help reuse of code a little? I'd especially avoid having to copy paste the report name constants everywhere.

e.g. change `buildOutput.assertValueWithinRange(EXPECTED_IMAGE_SIZE, TOLERANCE_PERCENTAGE, "image_details", "total_bytes");`
to `buildOutput.expectImageSize( long expectedzie, int tolerancePercentage)` ?

Sanne replied:
Also, I'm wondering if you really want to allow some margin of error for each of these. For example, does it make sense to have tolerante about EXPECTED_JNI_TYPES ?
With a dedicated method you'd be able to nudge the other developers in the right direction.

zakkak replied:
> Perhaps it could be reworked to help reuse of code a little? 

Done, it's much cleaner now and uses properties files for the expected value and the tolerance instead of code. This way it's more flexible and doesn't require code changes to add a new check.

> For example, does it make sense to have tolerante about EXPECTED_JNI_TYPES ?

Sure, for some cases we can be stricter. Happy to adjust the tolerance to whatever the Quarkus team would feel comfortable with (note that the stricter the checks the more the expected test failures in the CI).

galderz commented:
LGTM so far @zakkak

jerboaa commented on integration-tests/main/src/test/resources/image-metrics-test.properties:
```diff
@@ -0,0 +1,23 @@
+image_details.total_bytes=138778008
+image_details.total_bytes.tolerance=3
+# TODO: Switch to using analysis_results.types.total key once we drop support for GraalVM 22.3.0
```
This comment seems outdated? Did you mean `analysis_results.types.reachable`?

zakkak replied:
Correct

jerboaa commented on integration-tests/main/src/test/resources/image-metrics-test.properties:
```diff
@@ -0,0 +1,23 @@
+analysis_results.methods.reachable.tolerance=3
+analysis_results.fields.reachable=44161
+analysis_results.fields.reachable.tolerance=3
+# TODO: Switch to using analysis_results.types.total key once we drop support for GraalVM 22.3.0
```
`types.reflection`?

zakkak replied:
Correct

jerboaa commented on integration-tests/main/src/test/resources/image-metrics-test.properties:
```diff
@@ -0,0 +1,23 @@
+analysis_results.methods.reflection.tolerance=3
+analysis_results.fields.reflection=438
+analysis_results.fields.reflection.tolerance=3
+# TODO: Switch to using analysis_results.types.total key once we drop support for GraalVM 22.3.0
```
`types.jni`?

zakkak replied:
Correct

jerboaa reviewed (approved):
OK. Just some minor comments.

zakkak committed:
Add support for different thresholds per GraalVM version

`ImageMetricsITCase` tests are skipped for versions with undefined
thresholds

zakkak commented:
I updated the thresholds and added support for different thresholds per Mandrel/GraalVM version. @jerboaa please review https://github.com/quarkusio/quarkus/pull/36108/commits/8392e76a5526a6491c71df39df4862576f6ab52c

zakkak commented:
> ImageMetricsITCase.verifyImageMetrics:15 Expected analysis_results.methods.reflection to be within range [7472 +- 3%] but was 7734 ==> expected: <true> but was: <false>

That's interesting... The test fails even after updating the thresholds...

I am running the test locally and the number of methods registered for reflection are indeed close to 7470 (+-4 methods), while on the github runner they appear to be higher with no apparent reason.

It's true that the compilation process is not deterministic but that's a big difference :/

In fact all metrics seem off

```

zakkak commented:
> with no apparent reason.

It turns out I was running without `-Dtest-containers`... I will update the thresholds once more

zakkak commented:
Test failures are irrelevant (see https://github.com/quarkusio/quarkus/pull/37347#issuecomment-1831577444)

jerboaa reviewed (approved):
Seems fine to me.

maxandersen commented on integration-tests/jpa-postgresql-withxml/src/test/java/io/quarkus/it/jpa/postgresql/ImageMetricsITCase.java:
```diff
@@ -0,0 +1,17 @@
+
+@ExtendWith(BuildOutputExtension.class)
+@QuarkusIntegrationTest
+public class ImageMetricsITCase {
```
do I get it right that idea is to just add a test of these 17 characters in every IT test case and since the general structure is IT tests will include all the various extensions it will automagically works - no extra setup?

zakkak replied:
Yes, if you follow the three steps mentioned in the description of this PR you should be good to go:

> 1. copying ImageMetricsITCase.java in the corresponding folder
> 2. running the test once
> 3. running the following script passing the build output json file as the first parameter.

geoand commented:
It would be nice if there was a writeup of what the state of the PR is after all the changes that went into it, as it would help people reviewing for the first time and free them from the burden of having to build up the context by reading all the comments.

Thanks!

zakkak commented:
@geoand the description of the PR is up to date, there is no need to go through all the comments.

geoand commented:
👍🏼

geoand commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/nativeimage/BuildOutputExtension.java:
```diff
@@ -0,0 +1,138 @@
+ * <a href="https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md">the upstream GraalVM
+ * documentation</a>.
+ */
+public class BuildOutputExtension implements BeforeAllCallback {
```
Minor nitpick - can we improve the name of this? Perhaps `NativeBuildOutputExtension`?

zakkak replied:
Done

zakkak committed:
Rename `BuildOutputExtension` to `NativeBuildOutputExtension`

geoand reviewed (approved):
LGTM

zakkak commented:
I rebased the branch to have a more recent CI run. If things look good and there are no objections I plan to merge this tonight.

jerboaa commented:
@zakkak We see some failures related to that on GraalVM master built with JDK 21 here:

```
Error:  Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.406 s <<< FAILURE! -- in io.quarkus.it.jpa.postgresql.ImageMetricsITCase
Error:  io.quarkus.it.jpa.postgresql.ImageMetricsITCase.verifyImageMetrics -- Time elapsed: 0.332 s <<< FAILURE!
org.opentest4j.AssertionFailedError: Expected analysis_results.methods.reachable to be within range [96465 +- 3%] but was 104006 ==> expected: <true> but was: <false>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:214)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.assertValueWithinRange(NativeBuildOutputExtension.java:83)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.lambda$verifyImageMetrics$0(NativeBuildOutputExtension.java:59)
	at java.base/java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1603)
	at java.base/java.util.Properties.forEach(Properties.java:1422)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.verifyImageMetrics(NativeBuildOutputExtension.java:52)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.verifyImageMetrics(NativeBuildOutputExtension.java:46)
	at io.quarkus.it.jpa.postgresql.ImageMetricsITCase.verifyImageMetrics(ImageMetricsITCase.java:16)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

https://github.com/graalvm/mandrel/actions/runs/7228893522/job/19699465544#step:12:518

jerboaa commented:
Also for Mandrel 23.1 for the `main` integration test:

```
Error:  io.quarkus.it.main.ImageMetricsITCase.verifyImageMetrics -- Time elapsed: 0.016 s <<< FAILURE!
org.opentest4j.AssertionFailedError: Expected analysis_results.fields.jni to be within range [61 +- 1%] but was 62 ==> expected: <true> but was: <false>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:214)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.assertValueWithinRange(NativeBuildOutputExtension.java:83)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.lambda$verifyImageMetrics$0(NativeBuildOutputExtension.java:59)
	at java.base/java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1603)
	at java.base/java.util.Properties.forEach(Properties.java:1422)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.verifyImageMetrics(NativeBuildOutputExtension.java:52)
	at io.quarkus.test.junit.nativeimage.NativeBuildOutputExtension.verifyImageMetrics(NativeBuildOutputExtension.java:46)
	at io.quarkus.it.main.ImageMetricsITCase.verifyImageMetrics(ImageMetricsITCase.java:15)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

https://github.com/graalvm/mandrel/actions/runs/7228893522/job/19699470818#step:12:619

zakkak commented:
In Mandrel 23.1 it looks like the new field being accessed through JNI is `sun.nio.fs.UnixFileAttributes.st_birthtime_sec` introduced with https://bugs.openjdk.org/browse/JDK-8316304

I will just increase the value (in anticipation of the January release) along with the percentage threshold (for backwards compatibility).

Didn't look into the GraalVM + `main` increase in reachable methods.

jerboaa commented:
> Since upstream GraalVM doesn't plan on releasing GraalVM 24.0 based on JDK 21 it's not expected to get a jvmci version bump in LabsJDK 21. Unfortunately I don't see any way to work around this. Perhaps we could make these tests to be opt-out, so that we can disable them in certain environments. WDYT?

+1 to an opt-out. In certain runs we should just be able to configure CI to not run the metrics tests. Environment variable or property comes to mind.

zakkak commented:
Opt-out option implemented in https://github.com/quarkusio/quarkus/pull/37820
